### PR TITLE
RUN-707: Use-Client-Side-FileReader-API-Load-Content-From-File

### PR DESCRIPTION
**Pull request by ehe-pd**
**Is this a bugfix, or an enhancement? Please describe.**
- Fixes RUN-707. It's an security enhancement by obsolete file upload feature for import storage key. 

**Describe the solution you've implemented**
- The uploading file feature is replaced by read the file content in the client-side through FileReader API so there is no file uploaded to the server. 

**Describe alternatives you've considered**
- We considered obsolete the file upload feature completely. But it will change the user experience significantly. To avoid confusing user we choose the way with minimum impact to the user experience.

**Additional context**
There is no noticeable UI change.

### CODE REVIEW

**Review by gschueler — COMMENTED** [review]
We will need to change the `KeyStorageSelector.vue` UI as well

**Review by gschueler — COMMENTED** [review]
* When using the File Upload button, the file name cannot be changed, the original file name is always used
* Selecting a non-text file (e.g. a PNG file) will fill the textarea with binary data. (The saved file seems to be corrupted and does not match the original file when downloaded.)

**Comment by ehe-pd** [comment]
> * When using the File Upload button, the file name cannot be changed, the original file name is always used
It follows the original design when using file upload, the filename is not editable. I will change the field to make it editable.

> * Selecting a non-text file (e.g. a PNG file) will fill the textarea with binary data. (The saved file seems to be corrupted and does not match the original file when downloaded.)
The binary file is broken because the content is always read as ASCII text.  If a user wants to use an arbitrary file as a key, this should be an expected result. This is a user experience topic, in my opinion, we can provide a text hint to inform users to provide a correct text file instead of accommodating anything they choose to upload.

**Comment by ehe-pd** [comment]
@gschueler 
> We will need to change the `KeyStorageSelector.vue` UI as well

I have tested the feature in Node Wizard and checked the source code of the component, this component has already used the same client-side way to upload keys instead of uploading a file. 

```
handleFileUpload(e: any){
                var files = e.target.files || e.dataTransfer.files;
                if (!files.length)
                    return;

                const file = files[0];
                this.upload.file = file.name;

                let reader = new FileReader();
                reader.onload = (event:any) => {
                    let text = event.target.result;
                    this.upload.fileContent = text;
                    if(this.upload.errorMsg!=null){
                        this.upload.errorMsg = null;
                    }
                };
                reader.onerror = (event:any) => {
                    this.upload.errorMsg = "file cannot be read";
                    this.upload.file = null;
                };
                reader.readAsText(file);

            },
```

File: rundeckapp/grails-app/controllers/rundeck/controllers/StorageController.groovy [hunk 1]
```diff
@@ -268,19 +269,13 @@ class StorageController extends ControllerBase{
...
-        }.invalidToken{
+
+        withForm { }.invalidToken{
             flash.errorCode= 'request.error.invalidtoken.message'
             return redirect(controller: 'menu', action: 'storage', params: [project: params.project])
         }
-        if(!valid){
-            return
-        }
+
```
Comment (gschueler) [inline rundeckapp/grails-app/controllers/rundeck/controllers/StorageController.groovy:277]: the `valid` variable is still necessary. without that check, execution of `keyStorageUpload` method will not return.  the `return` inside the closure does not return from the outer method

**Review by gschueler — APPROVED** [review]

**Pull request merged**